File: backend/src/main/kotlin/com/example/backend/services/TicketService.kt
package com.example.backend.services

import com.example.backend.models.TicketCaisse
import com.example.backend.repositories.TicketCaisseRepository
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

@Service
class TicketCaisseService(
    private val ticketCaisseRepository: TicketCaisseRepository
) {

    fun getAllTickets(): List<TicketCaisse> {
        return ticketCaisseRepository.findAll()
    }

    fun getTicketById(id: Int): TicketCaisse {
        return ticketCaisseRepository.findById(id)
            .orElseThrow { IllegalArgumentException("Ticket with ID $id not found") }
    }

    fun getTicketByCodebarre(codebarre: Int): TicketCaisse? {
        return ticketCaisseRepository.findByCodebarre(codebarre)
    }

fun createTicket(ticket: TicketCaisse): TicketCaisse {
    val dateCode = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyMMddHHmmss"))
    ticket.codebarre = dateCode.toIntOrNull() ?: throw IllegalArgumentException("Failed to generate codebarre")
    ticket.dateGenere = LocalDate.now()
    ticket.supprimer = 0
    return ticketCaisseRepository.save(ticket)
}

    fun updateTicket(id: Int, updatedTicket: TicketCaisse): TicketCaisse {
        val existingTicket = ticketCaisseRepository.findById(id)
            .orElseThrow { IllegalArgumentException("Ticket with ID $id not found") }

        existingTicket.codebarre = updatedTicket.codebarre
        existingTicket.montant = updatedTicket.montant
        existingTicket.statut = updatedTicket.statut
        existingTicket.validite = updatedTicket.validite
        return ticketCaisseRepository.save(existingTicket)
    }

    fun deleteTicket(id: Int) {
        val ticket = ticketCaisseRepository.findById(id)
            .orElseThrow { IllegalArgumentException("Ticket with ID $id not found") }
        ticket.supprimer = 1
        ticketCaisseRepository.save(ticket)
    }
}

File: backend_py/pharmaPython/app/services/ticket_service.py
# services/ticket_service.py
from sqlalchemy.orm import Session
from datetime import datetime, date

from app.repositories.ticket_caisse_repository import TicketCaisseRepository


class TicketService:
  def __init__(self, db: Session):
    self.db = db
    self.ticket_repo = TicketCaisseRepository(db)

  def get_all_tickets(self):
    return self.ticket_repo.find_all()

  def get_ticket_by_id(self, id_: int):
    t = self.ticket_repo.find_by_id(id_)
    if not t:
      raise ValueError(f"Ticket with ID {id_} not found")
    return t

  def get_ticket_by_codebarre(self, codebarre: int):
    return self.ticket_repo.find_by_codebarre(codebarre)

  def create_ticket(self, ticket):
    date_code = datetime.now().strftime("%y%m%d%H%M%S")
    ticket.codebarre = int(date_code)
    ticket.dateGenere = date.today()
    ticket.supprimer = 0
    return self.ticket_repo.save(ticket)

  def update_ticket(self, id_: int, updated_ticket):
    existing = self.ticket_repo.find_by_id(id_)
    if not existing:
      raise ValueError(f"Ticket with ID {id_} not found")
    existing.codebarre = updated_ticket.codebarre
    existing.montant = updated_ticket.montant
    existing.statut = updated_ticket.statut
    existing.validite = updated_ticket.validite
    return self.ticket_repo.save(existing)

  def delete_ticket(self, id_: int):
    t = self.ticket_repo.find_by_id(id_)
    if not t:
      raise ValueError(f"Ticket with ID {id_} not found")
    t.supprimer = 1
    return self.ticket_repo.save(t)
